jlm/llvm/opt/alias-analyses/OnlineCycleDetection.hpp

```diff
@@ -0,0 +1,390 @@
+
+    // Perform DFS to find all roots reachable from superset
+    // We only care about nodes where with its topo index <= upperTopo
+    // These nodes will all need to be moved after subset in the new topological order
+    JLM_ASSERT(DfsStack_.empty());
+    // All nodes that have been pushed to the dfs stack
+    DfsNodesVisited_.Clear();
+    // All nodes that could reach subset and thus are part of a subset -> superset -> subset cycle
+    DfsCycleNodes_.Clear();
+
+    // Start DFS from the superset
+    DfsStack_.emplace(superset, false);
+
+    while (!DfsStack_.empty())
+    {
+      const auto node = DfsStack_.top().first;
+      const bool firstVisit = !DfsStack_.top().second;
+      DfsStack_.pop();
+      JLM_ASSERT(Set_.IsUnificationRoot(node));
+
+      if (firstVisit)
+      {
+        // If a "firstVisit" of this node has already occurred, skip it.
+        // This can happen if the same node is pushed multiple times before being popped.
+        if (!DfsNodesVisited_.Insert(node))
+          continue;
+
+        // Push node again to visit it a second time on the way back
+        DfsStack_.emplace(node, true);
+
+        // Push all non-pushed successors, as long as their position in the topological order is
+        // between [lowerTopo, upperTopo]. Nodes outside this range can stay where they are in the
+        // topological order, and will never be part of a path from superset to subset.
+        for (auto successor : GetSuccessors_(node))
+        {
+          auto successorParent = Set_.GetUnificationRoot(successor);
+
+          // Only push nodes that have yet to be visited
+          if (DfsNodesVisited_.Contains(successorParent))
+            continue;
+
+          // Ensure that the topological order invariants are satisfied
+          JLM_ASSERT(ComesAfter(successorParent, node));
+
+          // Only care about nodes within the topological range currently being rearranged
+          const auto topo = ObjectToTopoOrder_[successorParent];
+          if (topo < lowerTopo || topo > upperTopo)
+            continue;
+
+          // Visit the successor
+          DfsStack_.emplace(successorParent, false);
+        }
+      }
+      else
+      {
+        // Node is being visited for the second time, on the way back in the dfs
+        // If this node is the subset, there is a cycle
+        if (node == subset)
+        {
+          DfsCycleNodes_.Insert(node);
+          continue;
+        }
+
+        // Check if any of node's successors reach subset
+        for (auto successor : GetSuccessors_(node))
+        {
+          auto successorParent = Set_.GetUnificationRoot(successor);
+          if (DfsCycleNodes_.Contains(successorParent))
+          {
+            DfsCycleNodes_.Insert(node);
+            break;
+          }
+        }
+      }
+    }
+
+    // This list is used to store all roots that should come after the subset in the new ordering
+    DfsSupersetAndBeyond_.clear();
```

Comment:
This method can also be pushed in the `ClearInternalState()` method, no? I do not see `DfsSupersetAndBeyond` used before this line.